### DISCUSSION TITLE
Interrupt on both edges of a switch

### DISCUSSION THREAD

**Discussion by DracoTomes** [micropython/micropython#13150] [General]
Hello all,

I wanna detect a physical switch being flipped on a pin with mode `Pin.OUT`.
From what I've researched it's impossible to have multiple `Pin.irq()` active at the same time, one for `Pin.IRQ_FALLING` and one for `Pin.IRQ_RISING`.
[In this old Reddit post ](https://www.reddit.com/r/esp32/comments/ia9bsa/comment/g1lxr83/?utm_source=share&utm_medium=web2x&context=3) someone got it working with `trigger=Pin.IRQ_RISING | Pin.IRQ_FALLING` but I couldn't get it to work on my pi pico with 1.20 thus far.
Should I just run a regular check in an asyncio coroutine?

**Comment by rkompass** [comment]
This works:
```py
from machine import Pin
from time import sleep_ms

p0 = Pin(0, Pin.OUT)

def p_isr(pin):
    print(f"pin {pin} is now {'high' if pin() else 'low'}")

p0.irq(trigger=Pin.IRQ_RISING | Pin.IRQ_FALLING, handler=p_isr)

p0(1)
sleep_ms(20)
p0(0)
sleep_ms(20)
p0(1)
sleep_ms(20)
p0(0)
```
Output:
```
pin Pin(GPIO0, mode=OUT) is now high
pin Pin(GPIO0, mode=OUT) is now low
pin Pin(GPIO0, mode=OUT) is now high
pin Pin(GPIO0, mode=OUT) is now low
```
And it should work the same if the pin mode is `Pin.IN`.
In your ISR you might check which flank did the triggering, if you want to treat the flanks differently.

**Reply by robert-hh** [reply]
@DracoTomes I had just set up a similar test, using Pin.IN and an external source for a square wave signal. Works as expected, triggering at both slopes of the signal. I wonder why you have set the Pin to `OUT` mode. For detecting the state of an external switch, `IN` mode would be appropriate. Also, when using a mechanical switch, you must deal with contact bouncing, which will generate a fast burst of on/off states during the transition.

**Reply by DracoTomes** [reply]
It turns out I've never actually written that part of the interrupt handler but assumed I had in my head.
The reason it's in `OUT` mode is because the pin switches a relay that is also controlled by a hardware override - if the external switch is pressed I would like to update the internal state to reflect that.
This works flawlessly now that the pico doesn't have to read my thoughts anymore.
